Add a lazy Linux repositories' synchronization service with different download strategies

waterflow80 opened the pull request:
## What does this PR change?
### Lzreposync
The purpose of this PR is to add a new command line utility named `lzreposync`, which lazily synchronizes linux (Debian & Redhat) packages' metadata.

Before this commit, Uyuni uses `spacewalk-repo-sync` as a tool to synchronize repositories and packages. The problem with this tool is that it fully downloads all the packages (metadata + binaries) of a given repository, which is very resource-consuming, and most of the imported packages will not be immediately required for use.

The `lzreposync` tool split the synchronization process into two
steps:
1. synchronize the repositories' metadata
2. download the actual packages (binares) later

This commit contains the implementation of the first step: the metadata synchronization.

There are a different set of metadata files for each distribution:
- For Debian: `packages` and `translation-xx`
- For Redhat: `repomd.xml`, `primary.xml`, `filelists.xml`, and
`updateinfo.xml`

The metadata file parsers are implemented in a way that optimizes memory usage, by doing 'incremental' parsing rather than loading the whole file in memory at once.

### Lzdownload
The `lzdownalod`  service will be in charge of using the cached packages' metadata to download the actual binaries (or source rpms).

In this PR, we implemented a minimal version of the `download all` strategy, by downloading all the packages of given  channel using their cached metadata.

#### Usage
```sh
pip install -e .
lzdownload --channel channel_label

# you should see something like
00:55:40 Downloading total 423 files from 1 queues.
00:55:40     1/423 : gstreamer-plugins-bad-1.22.0-lp155.3.10.1.x86_64.rpm
00:55:40     2/423 : gstreamer-plugins-bad-1.22.0-lp155.3.14.1.x86_64.rpm
...
```
And the packages will be downloaded to the specified location in the filesystem:
```sh
├── 5f41124267d1c6329289e83f06f8e33bced9862222b5a85f1e0dea92d763463a
│   └── gstreamer-plugins-bad-1.22.0-lp155.3.14.1.src.rpm
├── c7a4d7475745bb2c9b747a5d821e147031aca786e43c19948a5fff51a50df177
│   └── gstreamer-plugins-bad-1.22.0-lp155.3.4.1.src.rpm
├── f3e52634346fb67c6c92a74cd308124693edac848f0521fecdaf671e3ec9c209
│   └── gstreamer-plugins-bad-1.22.0-lp155.3.10.1.src.rpm
├── f690d6767df5e1ae68bb1442739b502d8f5baade733121e8ac98ceb34cfbbb0b
```

#### Independence
We have separated the `lzdownload` from the `lzreposync` so that each service can run independently from each other. This will help in scaling and in the separation of tasks.
We may consider putting some functions used by both services in a common location.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
ℹ️ If a major new functionality is added, it is **strongly recommended** that tests for the new functionality are added to the Cucumber test suite
- No tests: Unit tests will be added on the fly.

- [x] **DONE**

## Links

Issue(s): #
Port(s): # **add downstream PR(s), if any**

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

agraul committed:
Create lzreposync subdirectory

lzreposync will be a spacewalk-repo-sync replacement written in Python.
It uses a src layout and a pyproject.toml. The target Python version is
3.11, compatibility with older Python versions is explicitly not a goal.

waterflow80 committed:
Add loop functionality to service

Made the lzreposync service continuously loop over the existing
channels and synchronize the corresponding repositories.

Added a status column in the rhnchannel table to indicate the
sync status of a given channel.

Also added some helper arguments to the service that allows us to
perform test operations, like creating a test channel and
associating repositories to it, etc

waterflow80 committed:
Implement the download all strategy

Implemented the `download all strategy`, meaning that for a
given channel, we download all the packages that are linked to
that channel.

cbosdo commented:
@waterflow80 Just letting you know that I'm not forgetting to review your PR... just overloaded these days

sbluhm commented on python/lzreposync/src/lzreposync/primary_parser.py:
```diff
@@ -0,0 +1,386 @@
+    def __init__(self, primary_file, repository="", arch_filter=".*"):
+        """
+        primary_file: In gzip format
+        # TODO: use  uyuni.common.fileutils.decompress_open for different format handling (gz, xz, ect), However! we should close the file manually
```
Hello @waterflow80 , I just stumbled across this line. What is the reason why you didn't implement decompress_open in the first place? It looks like a drop-in replacement.

waterflow80 replied:
Hello :wave:,
Hmm, looking at it now, it seems that implementing `decompress_open`, with some adaptation, will do fine.

I believe the main point here was that `decompress_open` returns a reference to a `GzipFile` object (for gzip files) but it does not check if the file is corrupted or not, as the opening and decompression is lazily made. 

So, in `is_valid_primary_file` we tried to check if the file is not corrupted and if it is ready for parse, by reading the first byte (not sure if that's enough though).
And I think I had some time constraints back then so I didn't complete the implementation for all formats.

So, yeah I believe that we can re-implement `decompress_open` with some adaptation and use it here.

I hope that answers the question :grinning:

sbluhm replied:
Thank you! To me, it looks like it is the same (apart from the TextIOWrapper):
```
def decompress_open(filename):
    file_obj = None
    if filename.endswith(".gz"):
        file_obj = gzip.open(filename, "rb")
...
    if filename.endswith((".gz", ".bz2", ".xz", ".zck", ".zst")):
        return io.TextIOWrapper(file_obj, encoding="utf8")
    return file_obj
```
I probably don't fully understand the code 😀
Let me play around with it some.

waterflow80 committed:
Add lazy linux repositories' metadata synchronization

The purpose of this commit is to add a new command line utility
named `lzreposync`, which lazily synchronizes linux (Debian &
Redhat) packages' metadata.

Before this commit, Uyuni uses `spacewalk-repo-sync` as a tool to
synchronize repositories and packages. The problem with this tool
is that it fully downloads all the packages (metadata + binaries)
of a given repository, which is very resource-consuming, and most
of the imported packages will not be immediately required for use.

The `lzreposync` tool split the synchronization process into two
steps:
1. synchronize the repositories' metadata
2. download the actual packages (binares) later

This commit contains the implementation of the first step: the
metadata synchronization.

There are a different set of metadata files for each distribution:
- For Debian: `packages` and `translation-xx`
- For Redhat: `repomd.xml`, `primary.xml`, `filelists.xml`, and
`updateinfo.xml`

The metadata file parsers are implemented in a way that optimizes
memory usage, by doing 'incremental' parsing rather than loading
the whole file in memory at once.
